### PR TITLE
Sync audit formal verification

### PR DESCRIPTION
**Pull request by brittcyr**
Synced to head

Known issues:
idl generation broken
ts tests broken

Have not verified if CI for FV works, but wont until this is merged

### CODE REVIEW

**Comment by brittcyr** [comment]
Fixed red black tree and removed rules that are expected to fail

**Comment by brittcyr** [comment]
Shank now fails because
Error: Rust Custom Composite IDL type not yet supported

File: client/rust/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ test = []
 anyhow = { workspace = true }
 manifest = { path = "../../programs/manifest", features=['no-clock'] }
 hypertree = { path = "../../lib" }
-jupiter-amm-interface = "0.4.5"
+jupiter-amm-interface = "=0.4.5"
```
Comment (DonDuala) [inline client/rust/Cargo.toml:17]: typo?

File: lib/Cargo.toml [hunk 2]
```diff
@@ -13,10 +13,18 @@ fuzz = []
 trace = []
 test = []
 
+# Certora feature exposes many of the internal implementation details, so should
+# only be used in formal verification. Ironically, the security testing makes
+# the code less secure.
```
Comment (DonDuala) [inline lib/Cargo.toml:18]: make note about conditional compiling even if obvs to any dev

File: programs/manifest/src/program/instruction_builders/batch_update_instruction.rs [hunk 3]
```diff
@@ -1,12 +1,19 @@
+#[cfg(feature = "certora")]
```
Comment (DonDuala) [inline programs/manifest/src/program/instruction_builders/batch_update_instruction.rs:1]: Is there any alternative to having these all over the place now?

**Review by DonDuala — APPROVED** [review]

File: client/rust/Cargo.toml (hunk 1 above)
Comment (brittcyr) [reply client/rust/Cargo.toml:17]: Intentional

File: programs/manifest/src/program/instruction_builders/batch_update_instruction.rs (hunk 3 above)
Comment (brittcyr) [reply programs/manifest/src/program/instruction_builders/batch_update_instruction.rs:1]: I got rid of a bunch of them.
Might be able to get rid of a few more if we are alright expanding scope of some functions, but dont really want to do that

File: lib/Cargo.toml (hunk 2 above)
Comment (brittcyr) [reply lib/Cargo.toml:18]: I can mention something in a readme. That comment is what a feature is in a Cargo.toml

Comment (DonDuala) [reply lib/Cargo.toml:18]: Ya think something for the uninformed on formal verification generally in the readme is useful

**Pull request merged**